<a href="https://colab.research.google.com/github/SachinScaler/DAV1Aug24/blob/main/Pandas_5%7CLecture_Dealing_with_NaNs%2C_String_and_Datetime.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!gdown 173A59xh2mnpmljCCB9bhC4C5eP2IS6qZ
import pandas as pd
import numpy as np

data = pd.read_csv('Pfizer_1.csv')

data_melt = pd.melt(data,id_vars = ['Date', 'Drug_Name', 'Parameter'],
            var_name = "time",
            value_name = 'reading')

data_tidy = data_melt.pivot(index=['Date','time', 'Drug_Name'],
                                        columns = 'Parameter',
                                        values='reading')
data_tidy = data_tidy.reset_index()
data_tidy.columns.name = None

Downloading...
From: https://drive.google.com/uc?id=173A59xh2mnpmljCCB9bhC4C5eP2IS6qZ
To: /content/Pfizer_1.csv
100% 1.51k/1.51k [00:00<00:00, 4.51MB/s]


In [2]:
data_tidy.head()

Date      time                Drug_Name  Pressure  Temperature
0  15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1  15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2  15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3  15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4  15-10-2020  11:30:00      docetaxel injection      29.0         25.0

In [3]:
data_tidy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Date         108 non-null    object 
 1   time         108 non-null    object 
 2   Drug_Name    108 non-null    object 
 3   Pressure     95 non-null     float64
 4   Temperature  95 non-null     float64
dtypes: float64(2), object(3)
memory usage: 4.3+ KB


In [13]:
a = None
print(type(a))

<class 'NoneType'>


In [14]:
import numpy as np

b = np.nan
print(type(b))

<class 'float'>


In [15]:
pd.Series([1, 2, np.nan])

0    1.0
1    2.0
2    NaN
dtype: float64

In [16]:
pd.Series([1, 2, 3])

0    1
1    2
2    3
dtype: int64

In [17]:
pd.Series([1, 2, 3.0])

0    1.0
1    2.0
2    3.0
dtype: float64

In [18]:
pd.Series([1, 2, np.nan, None])

0    1.0
1    2.0
2    NaN
3    NaN
dtype: float64

In [19]:
pd.Series(["1", "2", None])

0       1
1       2
2    None
dtype: object

In [20]:
pd.Series(["1", "2", None, np.nan])

0       1
1       2
2    None
3     NaN
dtype: object

### Detecting Null Values in DF and Column

In [21]:
num_series = pd.Series([1, 2, np.nan, None])

In [22]:
num_series.isna()

0    False
1    False
2     True
3     True
dtype: bool

In [23]:
num_series.isnull()

0    False
1    False
2     True
3     True
dtype: bool

In [25]:
obj_series = pd.Series(["1", "2", None, np.nan])

In [26]:
obj_series.isna()

0    False
1    False
2     True
3     True
dtype: bool

In [27]:
obj_series.isnull()

0    False
1    False
2     True
3     True
dtype: bool

In [28]:
obj_series.isnull().value_counts()

False    2
True     2
Name: count, dtype: int64

In [29]:
obj_series.isnull().sum() # count of Trues i.e Missing Values

2

Missing Vals on Df

In [30]:
data_tidy.head()

Date      time                Drug_Name  Pressure  Temperature
0  15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1  15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2  15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3  15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4  15-10-2020  11:30:00      docetaxel injection      29.0         25.0

In [32]:
data_tidy['Pressure'].isna().sum()

13

In [33]:
data_tidy.isna()

Date   time  Drug_Name  Pressure  Temperature
0    False  False      False     False        False
1    False  False      False     False        False
2    False  False      False     False        False
3    False  False      False     False        False
4    False  False      False     False        False
..     ...    ...        ...       ...          ...
103  False  False      False     False        False
104  False  False      False     False        False
105  False  False      False     False        False
106  False  False      False     False        False
107  False  False      False     False        False

[108 rows x 5 columns]

In [34]:
# colwise sum
data_tidy.isna().sum()

Date            0
time            0
Drug_Name       0
Pressure       13
Temperature    13
dtype: int64

In [38]:
data_tidy.isna().sum()/len(data_tidy) * 100

Date            0.000000
time            0.000000
Drug_Name       0.000000
Pressure       12.037037
Temperature    12.037037
dtype: float64

In [41]:
len(data_tidy), data_tidy.shape[0]

(108, 108)

In [35]:
# rowwise count
data_tidy.isna().sum(axis = 1)

0      0
1      0
2      0
3      0
4      0
      ..
103    0
104    0
105    0
106    0
107    0
Length: 108, dtype: int64

#### Extract the rows with missing value in a new DF

In [36]:
pd.isna

<function pandas.core.dtypes.missing.isna(obj: 'object') -> 'bool | npt.NDArray[np.bool_] | NDFrame'>

In [37]:
pd.isnull

<function pandas.core.dtypes.missing.isna(obj: 'object') -> 'bool | npt.NDArray[np.bool_] | NDFrame'>

### Handle Data
- removing
- imputing/replacing

### Removing Missing Values

In [42]:
data_tidy.shape

(108, 5)

In [43]:
data_tidy.head()

Date      time                Drug_Name  Pressure  Temperature
0  15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1  15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2  15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3  15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4  15-10-2020  11:30:00      docetaxel injection      29.0         25.0

In [51]:
data_drop_axis0 = data_tidy.dropna() # rows with missing values are dropped
data_drop_axis0

Date      time                Drug_Name  Pressure  Temperature
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0
..          ...       ...                      ...       ...          ...
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0

[95 rows x 5 columns]

In [46]:
data_drop_axis0.shape

(95, 5)

In [47]:
data_drop_axis1 = data_tidy.dropna(axis = 1) # cols with missing values are dropped
data_drop_axis1

Date      time                Drug_Name
0    15-10-2020  10:30:00  diltiazem hydrochloride
1    15-10-2020  10:30:00      docetaxel injection
2    15-10-2020  10:30:00   ketamine hydrochloride
3    15-10-2020  11:30:00  diltiazem hydrochloride
4    15-10-2020  11:30:00      docetaxel injection
..          ...       ...                      ...
103  17-10-2020   8:30:00      docetaxel injection
104  17-10-2020   8:30:00   ketamine hydrochloride
105  17-10-2020   9:30:00  diltiazem hydrochloride
106  17-10-2020   9:30:00      docetaxel injection
107  17-10-2020   9:30:00   ketamine hydrochloride

[108 rows x 3 columns]

In [49]:
data_drop_axis1.shape

(108, 3)

issue with dropping:
 - lots of valuable info might get deleted

### Imputing Data
#### Handling Missing Values

In [53]:
data_0 = data_tidy.fillna(0)
data_0.isna().sum()

Date           0
time           0
Drug_Name      0
Pressure       0
Temperature    0
dtype: int64

In [56]:
data_tidy['Pressure'].isna().sum()

13

In [57]:
data_tidy['Temperature'].isna().sum()

13

In [60]:
data_tidy['Pressure'].fillna(0)

0      18.0
1      26.0
2       9.0
3      19.0
4      29.0
       ... 
103    26.0
104    11.0
105     9.0
106    27.0
107    12.0
Name: Pressure, Length: 108, dtype: float64

In [61]:
data_tidy['Temperature'].fillna(-9999999999)

0      20.0
1      23.0
2      22.0
3      20.0
4      25.0
       ... 
103    19.0
104    20.0
105    13.0
106    20.0
107    21.0
Name: Temperature, Length: 108, dtype: float64

## common imputation terms:
 - fixed values

 ### statistical terms
 - mean: when data is equally distributed with no Outliers
 - median: When data has Outliers
 - mode: Categorical data


### Read Doc of Fillna and replace values with mean, median and mode for pressure

In [62]:
#resume@8:22

In [66]:
data_tidy['Pressure'].mean()

17.568421052631578

In [67]:
data_tidy['Pressure'].median()

17.0

In [87]:
data_tidy['Pressure'].mode()

0    22.0
Name: Pressure, dtype: float64

In [91]:
data_tidy['Pressure'].mode().to_list() # series to list

[22.0]

In [92]:
data_tidy['Pressure'].mode().values # series to array

array([22.])

In [82]:
data_tidy['Pressure'].mode().to_list()[0]

22.0

In [83]:
data_tidy['Pressure'].mode().values[0]

22.0

# values on series extract the values as numpy array

In [80]:
data_tidy['Pressure'].values

array([18., 26.,  9., 19., 29.,  9., 20., 28., 11., 12., nan,  8., 13.,
       22., nan, nan, 22., nan, 11., nan,  7., 13., 22., nan, 14., 23.,
        9., 16., nan, 10., 16., nan, 11., 24., 27., 10., 24., 28., 16.,
       nan, 29., 17., 27., 30., 18., 18., 23., 12., 19., 24., 12., 20.,
       nan, 13., 21., 25., nan, 22., 26., 15., 23., 27., 15., 24., 28.,
       15., 25., 29., 15., 25., 28., nan, 11., 28., 13., 13., 29., 14.,
       14., 28., 15.,  3., 20.,  8.,  4., 22.,  9.,  4., 22., 10.,  4.,
       22., 11.,  6., 22., 11.,  8., 23., 12.,  9., 25., 12., nan, 26.,
       11.,  9., 27., 12.])

In [85]:
mo =  data_tidy['Pressure'].mode().values[0]
data_tidy['Pressure'].fillna(mo)

0      18.0
1      26.0
2       9.0
3      19.0
4      29.0
       ... 
103    26.0
104    11.0
105     9.0
106    27.0
107    12.0
Name: Pressure, Length: 108, dtype: float64

In [86]:
mask = data_tidy['Pressure'].isna()

data_tidy['Pressure'].fillna(mo)[mask]

10     22.0
14     22.0
15     22.0
17     22.0
19     22.0
23     22.0
28     22.0
31     22.0
39     22.0
52     22.0
56     22.0
71     22.0
102    22.0
Name: Pressure, dtype: float64

In [94]:
data_tidy['Pressure'].fillna(mo).isna().sum()

0

In [95]:
sample = pd.Series(['1', '2', '3', np.NaN, None])
sample.fillna(0)

0    1
1    2
2    3
3    0
4    0
dtype: object

In [97]:
sample

0       1
1       2
2       3
3     NaN
4    None
dtype: object

In [98]:
# sample = sample.fillna(0)
sample.fillna(0, inplace = True)


In [99]:
sample

0    1
1    2
2    3
3    0
4    0
dtype: object

Based on the given DataFrame, which of the following statements regarding data imputation is mostly accurate?

|   CustomerID   |  TransactionAmount  |     Gender     |  Age  |  ProductCategory  |
|----------------|---------------------|----------------|-------|-------------------|
|       101      |         20          |     Male       |   35  |       Apparel     |
|       102      |        NaN          |     Female     |   28  |       NaN         |
|       103      |         15          |     Female     |  NaN  |       Electronics |
|       104      |         30          |     NaN        |   42  |       Electronics |
|       105      |        150          |     Male       |   30  |       Apparel     |
30 users have participated

- A: Imputing missing values in the "TransactionAmount" column using the mean of the available values may not be suitable due to potential skewness caused by outliers.
13%
- B: Imputing missing values in the "TransactionAmount" column using the median of the available values may be suitable to handle skewness due to outliers.
13%
- C: The presence of missing values in the "Gender" column can be effectively handled by imputing the most frequent category (mode).
10%
- D: All of the above

In [ ]:
# df['Gender'].fillna("Undisclosed")

### String

In [100]:
data_tidy

Date      time                Drug_Name  Pressure  Temperature
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0
..          ...       ...                      ...       ...          ...
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0

[108 rows x 5 columns]

In [102]:
data_tidy['Drug_Name'].apply(lambda x : 'hydrochloride' in x)

0       True
1      False
2       True
3       True
4      False
       ...  
103    False
104     True
105     True
106    False
107     True
Name: Drug_Name, Length: 108, dtype: bool

In [104]:
data_tidy['Drug_Name'].str.contains('hydrochloride')

0       True
1      False
2       True
3       True
4      False
       ...  
103    False
104     True
105     True
106    False
107     True
Name: Drug_Name, Length: 108, dtype: bool

In [111]:
sum([True, True, False])

2

In [103]:
data_tidy['Drug_Name']

0      diltiazem hydrochloride
1          docetaxel injection
2       ketamine hydrochloride
3      diltiazem hydrochloride
4          docetaxel injection
                ...           
103        docetaxel injection
104     ketamine hydrochloride
105    diltiazem hydrochloride
106        docetaxel injection
107     ketamine hydrochloride
Name: Drug_Name, Length: 108, dtype: object

In [105]:
data_tidy['Date']

0      15-10-2020
1      15-10-2020
2      15-10-2020
3      15-10-2020
4      15-10-2020
          ...    
103    17-10-2020
104    17-10-2020
105    17-10-2020
106    17-10-2020
107    17-10-2020
Name: Date, Length: 108, dtype: object

In [106]:
data_tidy['Date'].str.split('-')

0      [15, 10, 2020]
1      [15, 10, 2020]
2      [15, 10, 2020]
3      [15, 10, 2020]
4      [15, 10, 2020]
            ...      
103    [17, 10, 2020]
104    [17, 10, 2020]
105    [17, 10, 2020]
106    [17, 10, 2020]
107    [17, 10, 2020]
Name: Date, Length: 108, dtype: object

In [114]:
data_tidy['extractedDate'] = data_tidy['Date'].str.split('-').apply(lambda x: x[1])
data_tidy['extractedDate']

0      10
1      10
2      10
3      10
4      10
       ..
103    10
104    10
105    10
106    10
107    10
Name: extractedDate, Length: 108, dtype: object

In [ ]:
15-10-2020 -> 15, 10, 2020


In [116]:
data_tidy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Date           108 non-null    object 
 1   time           108 non-null    object 
 2   Drug_Name      108 non-null    object 
 3   Pressure       95 non-null     float64
 4   Temperature    95 non-null     float64
 5   extractedYear  108 non-null    object 
 6   extractedDate  108 non-null    object 
dtypes: float64(2), object(5)
memory usage: 6.0+ KB


In [118]:
data_tidy['ConvertedDate'] = pd.to_datetime(data_tidy['Date'])
data_tidy['ConvertedDate']

<ipython-input-118-8a05f6c1c91e>:1: UserWarning: Parsing dates in %d-%m-%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data_tidy['ConvertedDate'] = pd.to_datetime(data_tidy['Date'])


0     2020-10-15
1     2020-10-15
2     2020-10-15
3     2020-10-15
4     2020-10-15
         ...    
103   2020-10-17
104   2020-10-17
105   2020-10-17
106   2020-10-17
107   2020-10-17
Name: ConvertedDate, Length: 108, dtype: datetime64[ns]

In [119]:
data_tidy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date           108 non-null    object        
 1   time           108 non-null    object        
 2   Drug_Name      108 non-null    object        
 3   Pressure       95 non-null     float64       
 4   Temperature    95 non-null     float64       
 5   extractedYear  108 non-null    object        
 6   extractedDate  108 non-null    object        
 7   ConvertedDate  108 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(2), object(5)
memory usage: 6.9+ KB


In [120]:
data_tidy['Date'].dt.year

AttributeError: Can only use .dt accessor with datetimelike values

In [121]:
data_tidy['ConvertedDate'].dt.year

0      2020
1      2020
2      2020
3      2020
4      2020
       ... 
103    2020
104    2020
105    2020
106    2020
107    2020
Name: ConvertedDate, Length: 108, dtype: int32

In [122]:
data_tidy['ConvertedDate'].dt.month

0      10
1      10
2      10
3      10
4      10
       ..
103    10
104    10
105    10
106    10
107    10
Name: ConvertedDate, Length: 108, dtype: int32

### Can we convert full time stamp into Datetime

In [123]:
data_tidy

Date      time                Drug_Name  Pressure  Temperature  \
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0   
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0   
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0   
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0   
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0   
..          ...       ...                      ...       ...          ...   
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0   
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0   
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0   
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0   
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0   

    extractedYear extractedDate ConvertedDate  
0              10            10    2020-10-15  
1              10            10    2020-10-15  
2              10            10    2020-10-15  
3              10            10    2020-10-15  
4              10            10    2020-10-15  
..            ...           ...           ...  
103            10            10    2020-10-17  
104            10            10    2020-10-17  
105            10            10    2020-10-17  
106            10            10    2020-10-17  
107            10            10    2020-10-17  

[108 rows x 8 columns]

In [124]:
data['timestamp'] = data_tidy['Date'] + " " + data_tidy['time']
data['timestamp']

0     15-10-2020 10:30:00
1     15-10-2020 10:30:00
2     15-10-2020 10:30:00
3     15-10-2020 11:30:00
4     15-10-2020 11:30:00
5     15-10-2020 11:30:00
6     15-10-2020 12:30:00
7     15-10-2020 12:30:00
8     15-10-2020 12:30:00
9      15-10-2020 1:30:00
10     15-10-2020 1:30:00
11     15-10-2020 1:30:00
12     15-10-2020 2:30:00
13     15-10-2020 2:30:00
14     15-10-2020 2:30:00
15     15-10-2020 3:30:00
16     15-10-2020 3:30:00
17     15-10-2020 3:30:00
Name: timestamp, dtype: object

In [126]:
data['timestamp'] = pd.to_datetime(data['timestamp'])
data['timestamp']

<ipython-input-126-6e6f33f2edc6>:1: UserWarning: Parsing dates in %d-%m-%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  data['timestamp'] = pd.to_datetime(data['timestamp'])


0    2020-10-15 10:30:00
1    2020-10-15 10:30:00
2    2020-10-15 10:30:00
3    2020-10-15 11:30:00
4    2020-10-15 11:30:00
5    2020-10-15 11:30:00
6    2020-10-15 12:30:00
7    2020-10-15 12:30:00
8    2020-10-15 12:30:00
9    2020-10-15 01:30:00
10   2020-10-15 01:30:00
11   2020-10-15 01:30:00
12   2020-10-15 02:30:00
13   2020-10-15 02:30:00
14   2020-10-15 02:30:00
15   2020-10-15 03:30:00
16   2020-10-15 03:30:00
17   2020-10-15 03:30:00
Name: timestamp, dtype: datetime64[ns]

In [128]:
data['timestamp'].dt.year

0     2020
1     2020
2     2020
3     2020
4     2020
5     2020
6     2020
7     2020
8     2020
9     2020
10    2020
11    2020
12    2020
13    2020
14    2020
15    2020
16    2020
17    2020
Name: timestamp, dtype: int32

### HW:
- comparison
- sorting
- subtration and addition

In [143]:
data['timestamp'][0].strftime('%Y-%m-%d')

'2020-10-15'

In [ ]:
2020-10-15 -> 2020->Oct-15

In [146]:
df = pd.DataFrame([[1, '2020-01-01'], [2, '1998-01-12'], [3, '2012-11-05'],
                   [4, '2000-12-03'], [5, '1960-04-23'], [6, '2008-08-15']],
                    columns=["ID", "birth_dates"])

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ID           6 non-null      int64 
 1   birth_dates  6 non-null      object
dtypes: int64(1), object(1)
memory usage: 224.0+ bytes


In [147]:
df.head()

ID birth_dates
0   1  2020-01-01
1   2  1998-01-12
2   3  2012-11-05
3   4  2000-12-03
4   5  1960-04-23

In [150]:
df['birth_dates'] = pd.to_datetime(df['birth_dates'])
df['birth_dates']

0   2020-01-01
1   1998-01-12
2   2012-11-05
3   2000-12-03
4   1960-04-23
5   2008-08-15
Name: birth_dates, dtype: datetime64[ns]

In [151]:
df.iloc[2]['birth_dates'].year - df.iloc[1]['birth_dates'].year


14

### Write Dataframe into CSV

df.to_csv(pathtowriteat)

In [152]:
data_tidy.to_csv('pfizer_tidy.csv')

In [ ]:
#pd.read_csv(path)

HW: index=True what is the purpose of this parameter

In [4]:
df = pd.DataFrame({"SID": [1,1,1,2,2,2],
              "Sub": ["M", "P", "C","M", "P", "C", ],
              "Score": [85, 78, 92, 72, 80, 88]})
df

SID Sub  Score
0    1   M     85
1    1   P     78
2    1   C     92
3    2   M     72
4    2   P     80
5    2   C     88

In [5]:
import numpy as np
pd.pivot(index = "SID", columns = "Sub", data = df)

Score        
Sub     C   M   P
SID              
1      92  85  78
2      88  72  80

In [6]:
pd.melt(df, id_vars="SID", var_name="Sub")

SID    Sub value
0     1    Sub     M
1     1    Sub     P
2     1    Sub     C
3     2    Sub     M
4     2    Sub     P
5     2    Sub     C
6     1  Score    85
7     1  Score    78
8     1  Score    92
9     2  Score    72
10    2  Score    80
11    2  Score    88

In [7]:
pd.melt(df, id_vars="SID")

SID variable value
0     1      Sub     M
1     1      Sub     P
2     1      Sub     C
3     2      Sub     M
4     2      Sub     P
5     2      Sub     C
6     1    Score    85
7     1    Score    78
8     1    Score    92
9     2    Score    72
10    2    Score    80
11    2    Score    88

In [8]:
# Melting the dataset using 'Student_ID' as the identifier variable and
# 'Subject' as the variable to unpivot,
# then computing the mean of the 'Score' column.
pd.melt(df, id_vars="SID", var_name="Sub")

SID    Sub value
0     1    Sub     M
1     1    Sub     P
2     1    Sub     C
3     2    Sub     M
4     2    Sub     P
5     2    Sub     C
6     1  Score    85
7     1  Score    78
8     1  Score    92
9     2  Score    72
10    2  Score    80
11    2  Score    88

In [9]:
pd.melt(df, id_vars="SID")

SID variable value
0     1      Sub     M
1     1      Sub     P
2     1      Sub     C
3     2      Sub     M
4     2      Sub     P
5     2      Sub     C
6     1    Score    85
7     1    Score    78
8     1    Score    92
9     2    Score    72
10    2    Score    80
11    2    Score    88

In [10]:
df

SID Sub  Score
0    1   M     85
1    1   P     78
2    1   C     92
3    2   M     72
4    2   P     80
5    2   C     88

In [11]:
import pandas as pd

sales_data = {
    'Date': ['2024-01-01', '2024-01-01', '2024-01-02', '2024-01-02'],
    'Region': ['East', 'West', 'East', 'West'],
    'Product': ['A', 'B', 'A', 'B'],
    'Sales': [100, 200, 150, 250]
}

df = pd.DataFrame(sales_data)
df.head()


Date Region Product  Sales
0  2024-01-01   East       A    100
1  2024-01-01   West       B    200
2  2024-01-02   East       A    150
3  2024-01-02   West       B    250

In [12]:
pd.pivot_table(df, index = 'Region', values = 'Sales' , aggfunc = sum)

<ipython-input-12-54782e0d0a98>:1: FutureWarning: The provided callable <built-in function sum> is currently using DataFrameGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  pd.pivot_table(df, index = 'Region', values = 'Sales' , aggfunc = sum)


Sales
Region       
East      250
West      450